sigstore/_utils.py

```diff
@@ -104,3 +104,24 @@ def split_certificate_chain(chain_pem: str) -> list[bytes]:
     certificate_chain = [(pem_header + c).encode() for c in certificate_chain]
 
     return certificate_chain
+
+
+def sha256_streaming(io: IO[bytes]) -> bytes:
+    """
+    Compute the SHA256 of a stream.
+
+    This function does its own internal buffering, so an unbuffered stream
+    should be supplied for optimal performance.
+    """
+
+    sha256 = hashlib.sha256()
+    # Per coreutils' ioblksize.h: 128KB performs optimally across a range
+    # of systems in terms of minimizing syscall overhead.
+    view = memoryview(bytearray(128 * 1024))
+
+    nbytes = io.readinto(view)  # type: ignore
+    while nbytes:
+        sha256.update(view[:nbytes])
+        nbytes = io.readinto(view)  # type: ignore
```

Comment:
cc @znewman01, since you may have already run into this on `cosign` (or it may become a similar design consideration in the near future!)

Reply:
I think this is a bit outside our threat model, but if 2) is relatively easy, we should do that just to be safe. I have doubts on how well this will work cross-platform, though.

Reply:
Yeah, agreed. I'll leave it as-is for now, with a justifying comment.